File: src/chapters/nonlinear-equations/content/code/newton.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Newton's method using the derivative df.
double newton(function<double(double)> f, function<double(double)> df, double x0, double tol = 1e-12, int max_iter = 100) {
    double x = x0;
    for (int k = 0; k < max_iter; ++k) {
        double fx = f(x);
        if (fabs(fx) < tol) return x;
        x -= fx / df(x);
    }
    return x;
}

int main() {
    cout.precision(12);
    cout << newton([](double x) { return x * x - 2; }, [](double x) { return 2 * x; }, 1.0) << "\n";
}

File: src/chapters/nonlinear-equations/content/code/newton.rs
// Newton's method using f and its derivative df.
fn newton<F: Fn(f64) -> f64, D: Fn(f64) -> f64>(f: F, df: D, x0: f64, tol: f64, max_iter: usize) -> f64 {
    let mut x = x0;
    for _ in 0..max_iter {
        let fx = f(x);
        if fx.abs() < tol { return x; }
        x -= fx / df(x);
    }
    x
}
fn main() {
    println!("{}", newton(|x: f64| x * x - 2.0, |x: f64| 2.0 * x, 1.0, 1e-12, 100));
}
